nova/compute/manager.py

```diff
@@ -10680,6 +10681,35 @@ def _reclaim_queued_deletes(self, context):
                                 "instance: %s",
                                 e, instance=instance)
 
+    @periodic_task.periodic_task(
+            spacing=CONF.instances_stuck_in_deleting_cleanup_interval)
+    def _cleanup_instances_stuck_in_deleting(self, context):
+        """Reclaim instances stuck in DELETING state, but not really destroyed,
+        and retrigger terminate_instance for them.
+        """
+        filters = {'task_state': task_states.DELETING,
+                   'host': self.host}
+        instances = objects.InstanceList.get_by_filters(
+            context, filters,
+            expected_attrs=objects.instance.INSTANCE_DEFAULT_FIELDS,
+            use_slave=True)
+        for instance in instances:
+            # Check if the instance has a process lock
+            for name, s in _semaphores._semaphores.items():
+                if name == instance.uuid:
+                    break
+            else:
+                LOG.info(
+                        "Cleanup instance found in DELETING state but "
+                        "has no process lock", instance=instance)
+                try:
+                    bdms = objects.BlockDeviceMappingList.get_by_instance_uuid(
+                            context, instance.uuid)
+                    self.terminate_instance(context, instance, bdms)
```

Comment:
What if somebody took the instance lock between your check and your call to `self.terminate_instance()`? This looks pretty racy to me.

Reply:
Have you tried taking the lock with a timeout instead of checking the list? Maybe we can even make the lock re-entrant somehow so we can call `terminate_instance()` while still having the lock?

Reply:
if i understood synchronized() correctly, it will insure only one thread is running of on the instance, so if someone took the lock the synchronise  [here](https://github.com/sapcc/nova/blob/db78134420a4aa30c815d3fd3b465fd8aeb4d840/nova/compute/manager.py#L3329) will insure only one thread is running  and my call will wait until the lock got released by the first thread, if the instance got deleted by the first thread the exception catch [here](https://github.com/sapcc/nova/blob/db78134420a4aa30c815d3fd3b465fd8aeb4d840/nova/compute/manager.py#L3349) catches the exception, that the instance disappeared